dot/parachain/prospective-parachains/fragment-chain/fragment_chain.go

```diff
@@ -0,0 +1,1196 @@
+	}
+}
+
+func (c *CandidateStorage) AddPendingAvailabilityCandidate(
+	candidateHash parachaintypes.CandidateHash,
+	candidate parachaintypes.CommittedCandidateReceipt,
+	persistedValidationData parachaintypes.PersistedValidationData,
+) error {
+	entry, err := NewCandidateEntry(candidateHash, candidate, persistedValidationData, Backed)
+	if err != nil {
+		return err
+	}
+
+	return c.addCandidateEntry(entry)
+}
+
+// Len return the number of stored candidate
+func (c *CandidateStorage) Len() int {
+	return len(c.byCandidateHash)
+}
+
+func (c *CandidateStorage) addCandidateEntry(candidate *CandidateEntry) error {
+	_, ok := c.byCandidateHash[candidate.candidateHash]
+	if ok {
+		return ErrCandidateAlradyKnown
+	}
+
+	// updates the reference parent hash -> candidate
+	// we don't check the `ok` value since the key can
+	// exists in the map but pointing to a nil hashset
+	setOfCandidates := c.byParentHead[candidate.parentHeadDataHash]
+	if setOfCandidates == nil {
+		setOfCandidates = make(map[parachaintypes.CandidateHash]any)
+	}
+	setOfCandidates[candidate.candidateHash] = struct{}{}
+	c.byParentHead[candidate.parentHeadDataHash] = setOfCandidates
+
+	// udpates the reference output hash -> candidate
+	setOfCandidates = c.byOutputHead[candidate.outputHeadDataHash]
+	if setOfCandidates == nil {
+		setOfCandidates = make(map[parachaintypes.CandidateHash]any)
+	}
+	setOfCandidates[candidate.candidateHash] = struct{}{}
+	c.byOutputHead[candidate.outputHeadDataHash] = setOfCandidates
+
+	c.byCandidateHash[candidate.candidateHash] = candidate
+	return nil
+}
+
+func (c *CandidateStorage) removeCandidate(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		return
+	}
+
+	delete(c.byCandidateHash, candidateHash)
+
+	if setOfCandidates, ok := c.byParentHead[entry.parentHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byParentHead, entry.parentHeadDataHash)
+		}
+	}
+
+	if setOfCandidates, ok := c.byOutputHead[entry.outputHeadDataHash]; ok {
+		delete(setOfCandidates, candidateHash)
+		if len(setOfCandidates) == 0 {
+			delete(c.byOutputHead, entry.outputHeadDataHash)
+		}
+	}
+}
+
+func (c *CandidateStorage) markBacked(candidateHash parachaintypes.CandidateHash) {
+	entry, ok := c.byCandidateHash[candidateHash]
+	if !ok {
+		fmt.Println("candidate not found while marking as backed")
+	}
+
+	entry.state = Backed
+}
+
+func (c *CandidateStorage) contains(candidateHash parachaintypes.CandidateHash) bool {
+	_, ok := c.byCandidateHash[candidateHash]
+	return ok
+}
+
+// candidates returns an iterator over references to the stored candidates, in arbitrary order.
+func (c *CandidateStorage) candidates() iter.Seq[*CandidateEntry] {
+	return func(yield func(*CandidateEntry) bool) {
+		for _, entry := range c.byCandidateHash {
+			if !yield(entry) {
+				return
+			}
+		}
+	}
+}
```

Comment:
Can we return the slice instead?

Reply:
https://github.com/ChainSafe/gossamer/pull/4337#discussion_r1872041545

Reply:
Do we even need this method? It's called from `(f *FragmentChain) Unconnected()`, which is only used in tests, not in logic.
Please remove unnecessary methods. if you need this logic in test you can created test helper function.